### PR TITLE
Expand upper bounds on RangeBounds impls

### PR DESCRIPTION
**Pull request by kungfukennyg**
Per issue #64027 adds `?Sized` to all `RangeBounds<&T>` impls.

### CODE REVIEW

**Comment by kennytm** [comment]
This has broken type-inference of some existing tests.

```
2019-09-10T01:32:20.0705767Z ---- collections/btree/map.rs - collections::btree::map::BTreeMap<K, V>::range_mut (line 841) stdout ----
2019-09-10T01:32:20.0706521Z error[E0283]: type annotations required: cannot resolve `_: std::cmp::Ord`
2019-09-10T01:32:20.0707955Z   --> collections/btree/map.rs:848:25
2019-09-10T01:32:20.0708249Z    |
2019-09-10T01:32:20.0708436Z 10 | for (_, balance) in map.range_mut("B".."Cheryl") {
```

**Comment by kennethbgoodin** [comment]
Yup I'm looking into it. Is there an easy way to run that test case?

**Comment by kennytm** [comment]
r? @Mark-Simulacrum

I think we need a crater run but the queue is super long for now 😕

**Comment by Mark-Simulacrum** [comment]
Hm, yeah, inference breakage here is concerning.

@bors try to get a build for crater

**Comment by Mark-Simulacrum** [comment]
@craterbot run mode=check-only

**Comment by kennethbgoodin** [comment]
How long do crater runs typically take? I'm curious as to how much longer this PR may be in the queue. I see it's second to the top.

**Comment by Mark-Simulacrum** [comment]
I think the current average for a check run is around 4 days, but I'm not sure.

**Comment by jonas-schievink** [comment]
cc https://github.com/rust-lang/rust/pull/61584, an earlier attempt at this

**Comment by Centril** [comment]
Seems like acceptable breakage to me (a single crates.io regression which will probably be fixed soon...).

r? @dtolnay to assess this.

File: src/libcore/ops/range.rs [hunk 1]
```diff
@@ -904,7 +904,7 @@ impl<T> RangeBounds<T> for RangeTo<&T> {
 }
 
 #[stable(feature = "collections_range", since = "1.28.0")]
-impl<T> RangeBounds<T> for Range<&T> {
+impl<T: ?Sized> RangeBounds<T> for Range<&T> {
     fn start_bound(&self) -> Bound<&T> {
         Included(self.start)
     }
```
Comment (dtolnay) [inline src/libcore/ops/range.rs]: Was this change supposed to apply to the RangeBounds\<T\> impls for RangeInclusive\<&T\> and RangeToInclusive\<&T\> (below) as well?

**Review by dtolnay — COMMENTED** [review]
I agree that the `T: ?Sized` impls are more correct and it would be great to have that, but I am not thrilled about introducing even more need for turbofish on range and range_mut.

Some questions to dig into what is going on here:

1. Is there any defensible language change that might allow us to eliminate the ambiguity when using `"from".."to"` as a range into a BTreeMap\<&str, _\> or (separately) a BTreeMap\<String, _\>? Possibly some way to guide the inference of `T` when the other type parameters are known, or some way to say that any choice of `T` that satisfies the bounds is fine even if it's not known to be the unique possible choice.

2. If range and range_mut didn't exist yet, is there a better less generic signature for them that wouldn't be ambiguous in common cases?

I don't know who these questions are directed to but I am hesitant to merge this change as currently implemented. The amount of breakage is small and would probably be justifiable for a change that made `"from".."to"` "just work" without turbofishes, but as it is, I don't think landing this is the right tradeoff.

The workaround is to use `map.range::<str, _>((Bound::Included("from"), Bound::Excluded("to")))` which has the right performance characteristics, or `map.range("from".to_owned().."to".to_owned())` which is more concise.

**Comment by JohnTitor** [comment]
Ping from triage: @kennethbgoodin any updates on this?

**Comment by kennethbgoodin** [comment]
It sounds like the consensus is that this change isn't the best approach -- should I close this PR?

**Pull request closed without merging**